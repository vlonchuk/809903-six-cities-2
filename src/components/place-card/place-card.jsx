import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import {convertRatingToPercent} from './../../utils.js';
import PlacesListType from './../../consts/places-list-type.js';
import {Widths, Heights} from './../../consts/style.js';

class PlaceCard extends PureComponent {
  constructor(props) {
    super(props);

    this._handleMouseEnter = this.handleMouseEnter.bind(this);
    this._handleMouseLeave = this.handleMouseLeave.bind(this);
    this._handleAddToFavorite = this.handleAddToFavorite.bind(this);

    this._imgWidth = Widths.PROPERTY_IMAGE_LARGE;
    this._imgHeight = Heights.PROPERTY_IMAGE_LARGE;

    switch (this.props.listType) {
      case PlacesListType.CITY:
        this._cardClass = `cities__place-card place-card`;
        this._wrapperClass = `cities__image-wrapper place-card__image-wrapper`;
        this._infoClass = `place-card__info`;
        break;
      case PlacesListType.NEAR:
        this._cardClass = `near-places__card place-card`;
        this._wrapperClass = `near-places__image-wrapper place-card__image-wrapper`;
        this._infoClass = `place-card__info`;
        break;
      case PlacesListType.FAVORITES:
        this._cardClass = `favorites__card place-card`;
        this._wrapperClass = `favorites__image-wrapper place-card__image-wrapper`;
        this._infoClass = `favorites__card-info place-card__info`;
        this._imgWidth = Widths.PROPERTY_IMAGE_SMALL;
        this._imgHeight = Heights.PROPERTY_IMAGE_SMALL;
        break;
    }
  }

  handleMouseEnter() {
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(this.props.data);
    }
  }

  handleMouseLeave() {
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave();
    }
  }

  handleAddToFavorite() {
    this.props.onAddToFavorite(this.props.data.id, this.props.data.isFavorite ? 0 : 1, this.props.favorites);
  }

  render() {
    const bookmarkClass = `place-card__bookmark-button button` + (this.props.data.isFavorite ? ` place-card__bookmark-button--active` : ``);
    const roundedRating = Math.round(this.props.data.rating);
    const ratingWidth = convertRatingToPercent(roundedRating);

    return <article className={`${this._cardClass}`} id={this.props.data.id}
      onMouseEnter={this._handleMouseEnter} onMouseLeave={this._handleMouseLeave}>
      {
        this.props.data.isPremium ?
          <div className="place-card__mark">
            <span>Premium</span>
          </div>
          :
          null
      }

      <div className={`${this._wrapperClass}`}>
        <Link to={`/offer/${this.props.data.id}`}>
          <img className="place-card__image" src={this.props.data.previewImage} width={this._imgWidth} height={this._imgHeight}
            alt="Place image"></img>
        </Link>
      </div>
      <div className={this._infoClass}>
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">{`â‚¬`}{this.props.data.price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <button className={bookmarkClass} type="button" onClick={this._handleAddToFavorite}>
            <svg className="place-card__bookmark-icon" width={Widths.PLACE_CARD_BOOKMARK_ICON} height={Heights.PLACE_CARD_BOOKMARK_ICON}>
              <use xlinkHref="#icon-bookmark"></use>
            </svg>
            <span className="visually-hidden">To bookmarks</span>
          </button>
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: ratingWidth + `%`}}></span>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <Link to={`/offer/${this.props.data.id}`}>
          <h2 className="place-card__name">
            {this.props.data.title}
          </h2>
        </Link>
        <p className="place-card__type">{this.props.data.type}</p>
      </div>
    </article>;
  }
}

PlaceCard.propTypes = {
  listType: PropTypes.oneOf(Object.values(PlacesListType)).isRequired,
  data: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    isPremium: PropTypes.bool.isRequired,
    rating: PropTypes.number.isRequired,
  }),
  favorites: PropTypes.array.isRequired,
  onAddToFavorite: PropTypes.func.isRequired,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  favorites: state.favorites,
});

const PlaceCardWrapped = connect(mapStateToProps)(PlaceCard);
export {PlaceCard};
export default PlaceCardWrapped;
