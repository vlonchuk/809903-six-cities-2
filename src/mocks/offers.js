const offers = [
  {
    id: `prop-1`,
    city: `Amsterdam`,
    caption: `Beautiful & luxurious apartment at great location`,
    imgSrc: `img/apartment-01.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 120,
    priceText: `night`,
    rating: 50,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198
    },
  },
  {
    id: `prop-2`,
    city: `Amsterdam`,
    caption: `Wood and stone place`,
    imgSrc: `img/room.jpg`,
    type: `Private room`,
    priceCurrency: `€`,
    priceValue: 80,
    priceText: `night`,
    rating: 80,
    coor: {
      latitude: 52.369553943508,
      longitude: 4.85309666406198
    },
  },
  {
    id: `prop-3`,
    city: `Amsterdam`,
    caption: `Canal View Prinsengracht`,
    imgSrc: `img/apartment-02.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 132,
    priceText: `night`,
    rating: 100,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198
    },
  },
  {
    id: `prop-4`,
    city: `Amsterdam`,
    caption: `Nice, cozy, warm big bed apartment`,
    imgSrc: `img/apartment-03.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 180,
    priceText: `night`,
    rating: 65,
    coor: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198
    },
  },
  {
    id: `prop-5`,
    city: `Paris`,
    caption: `Canal View Prinsengracht`,
    imgSrc: `img/apartment-02.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 132,
    priceText: `night`,
    rating: 90,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198
    },
  },
  {
    id: `prop-6`,
    city: `Paris`,
    caption: `Nice, cozy, warm big bed apartment`,
    imgSrc: `img/apartment-03.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 180,
    priceText: `night`,
    rating: 40,
    coor: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198
    },
  },
];

export const getPropertiesByCity = (city) => {
  return offers.filter((el) => el.city === city)
               .slice();
};

export default offers;
