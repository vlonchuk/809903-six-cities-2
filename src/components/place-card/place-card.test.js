import React from 'react';
import renderer from 'react-test-renderer';
import {PlaceCard} from './place-card';
import offers from './../../mocks/offers.js';
import {BrowserRouter as Router} from 'react-router-dom';
import PlacesListType from './../../consts/places-list-type';

it(`PlaceCard correctly renders after relaunch`, () => {
  const item = offers[0];

  const onMouseEnter = jest.fn();
  const onMouseLeave = jest.fn();

  const tree = renderer
    .create(
        <Router>
          <PlaceCard key={item.id} data={item} onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave}
            listType={PlacesListType.CITY}
            favorites={[]}
            onAddToFavorite={jest.fn()}
          />
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
