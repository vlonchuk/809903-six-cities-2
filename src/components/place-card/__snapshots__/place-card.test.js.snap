// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceCard correctly renders after relaunch 1`] = `
<article
  className="cities__place-card place-card"
  id={1}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="/offer/1"
      onClick={[Function]}
    >
      <img
        alt="Place image"
        className="place-card__image"
        height={200}
        src="img/1.png"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          120
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height={19}
          width={18}
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <a
      href="/offer/1"
      onClick={[Function]}
    >
      <h2
        className="place-card__name"
      >
        Beautiful & luxurious studio at great location
      </h2>
    </a>
    <p
      className="place-card__type"
    >
      apartment
    </p>
  </div>
</article>
`;
